File: js/euler47.js
function euler47(){
  var streak = 0, n = 2;
  var primes = new Set();
  while (true){
    if (!isPrime(n)){
      // n is composite, count number of distinct prime factors
      let num = n, factors = new Set();
      for (let prime of primes){
        while (num > 1 && num % prime == 0){
          factors.add(prime);
          num /= prime;
        }
        if (num == 1) break;
      }
      if (factors.size == 4){
        streak++;
        if (streak == 4) return n - 3;
      }
      else streak = 0;
    }
    else{
      // n is prime, cache prime and reset streak 
      primes.add(n);
      streak = 0;
    }
    n++;
  }
}

function isPrime(n){
  for (let i = 2; i * i <= n; i++){
    if (n % i == 0) return false;
  }
  return true;
}

console.log(euler47());

File: go/euler47.go
package main
import "fmt"

func main() {
    isPrime := func(n int) bool {
        if n == 2 {
            return true
        }
        for i := 2; i * i <= n; i++ {
            if n % i == 0 {
                return false
            }
        }
        return true
    }
    
    curr, streak, res := 2, 0, 0
    primes := []int{}
    for res == 0 {
        if isPrime(curr) {
            primes = append(primes, curr)
            streak = 0
        } else {
            n, count := curr, 0
            factors := []int{}
            for n > 1 {
                for _, p := range primes {
                    if n % p == 0 {
                        count++
                        factors = append(factors, p)
                        for n % p == 0 {
                            n /= p
                        }
                    }
                }
            }
            if count == 4 {
                streak++
                if streak == 4 {
                    res = curr - streak + 1
                }
            } else {
                streak = 0
            }
        }
        curr++
    }
    fmt.Println(res)
}
